src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/DependencyInjectionEventSourceTests.cs

```diff
@@ -343,33 +331,77 @@ public void EmitsServiceProviderBuilt()
+            listener.EnableEvents(DependencyInjectionEventSource.Log, EventLevel.Verbose);
+            try
+            {
+                EventWrittenEventArgs serviceProviderBuiltEvent = listener.EventDataFor(provider).Single(e => e.EventName == "ServiceProviderBuilt");
+                Assert.Equal(1, GetProperty<int>(serviceProviderBuiltEvent, "singletonServices"));
 
-            _listener.EnableEvents(DependencyInjectionEventSource.Log, EventLevel.Verbose);
+                EventWrittenEventArgs serviceProviderDescriptorsEvent = listener.EventDataFor(provider).Single(e => e.EventName == "ServiceProviderDescriptors");
+                Assert.NotNull(JObject.Parse(GetProperty<string>(serviceProviderDescriptorsEvent, "descriptors")));
+            }
+            finally
+            {
+                listener.DisableEvents(DependencyInjectionEventSource.Log);
+            }
+        }
 
-            EventWrittenEventArgs serviceProviderBuiltEvent = _listener.EventData.Single(e => e.EventName == "ServiceProviderBuilt");
-            Assert.Equal(1, GetProperty<int>(serviceProviderBuiltEvent, "singletonServices"));
+        private static bool TryGetProperty<T>(EventWrittenEventArgs data, string propName, out T result)
+        {
+            if (data.PayloadNames is { } names &&
+                data.Payload is { } payload &&
+                names.IndexOf(propName) is var index and >= 0 &&
+                payload[index] is T value)
+            {
+                result = value;
+                return true;
+            }
 
-            EventWrittenEventArgs serviceProviderDescriptorsEvent = _listener.EventData.Single(e => e.EventName == "ServiceProviderDescriptors");
-            Assert.NotNull(JObject.Parse(GetProperty<string>(serviceProviderDescriptorsEvent, "descriptors")));
+            result = default;
+            return false;
         }
 
-        private T GetProperty<T>(EventWrittenEventArgs data, string propName)
-            => (T)data.Payload[data.PayloadNames.IndexOf(propName)];
+        private static T GetProperty<T>(EventWrittenEventArgs data, string propName) =>
+            TryGetProperty(data, propName, out T result)
+                ? result
+                : throw new ArgumentException($"No property {propName} with type {typeof(T)} found.");
+
+        public class TestEventListenerFixture : IDisposable
+        {
+            internal TestEventListener Listener { get; }
+
+            public TestEventListenerFixture()
+            {
+                Listener = new();
+                Listener.EnableEvents(DependencyInjectionEventSource.Log, EventLevel.Verbose);
+            }
+
+            public void Dispose()
+            {
+                Listener.DisableEvents(DependencyInjectionEventSource.Log);
+                Listener.Dispose();
+            }
+        }
 
-        private class TestEventListener : EventListener
+        internal class TestEventListener : EventListener
```

Comment:
why the change from private to internal here?

Reply:
Because the `TestEventListenerFixture` has a property of type `TestEventListener` that needs to visible in the test class.